### PR TITLE
Remove use of deprecated ClientAuthenticationMethod's

### PR DESCRIPTION
**Pull request by bibibiu2017**
Fixes gh-346

### CODE REVIEW

**Comment by sjohnr** [comment]
Hi @bibibiu2017, thanks for your interest in the project!

Unfortunately, this change has broken quite a lot of tests. Any chance you can take a look at those? Looking at the tests may highlight whether this alone is the correct change to make or not. I'm still catching up on the issue myself.

Also, please refer to [contributing guidelines](https://github.com/spring-projects-experimental/spring-authorization-server/blob/main/CONTRIBUTING.adoc) when submitting pull requests. In particular, you will want to update the header of files updated with the year "2021" in the copyright, squash your commits, and include "Fixes gh-346" on a separate line in your commit comment.

**Comment by bibibiu2017** [comment]
I have taken a look at the tests, the configured test registered clients, and some assertions are using the deprecated POST and BASIC Client authentication methods.

**Comment by sjohnr** [comment]
Thanks for looking at the tests, @bibibiu2017! I've tested on your branch, and while the immediate issue is resolved with this change, the sample project no longer works. This is because we've flipped the problem so now the deprecated client authentication methods are not allowed.

This is a difficult issue to solve for without too many changes, due to the fact that the `AuthenticationConverter`s being used aren't aware of the `RegisteredClient` which requests what method(s) should be supported, as that later runs into a conflict in [OAuth2ClientAuthenticationProvider.java#L103](https://github.com/spring-projects-experimental/spring-authorization-server/blob/687f03f047b693ee7ae6d885f46b63acefad5b86/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAuthenticationProvider.java#L103).

Of the available options, I think we can either:
1. Automatically add `CLIENT_SECRET_BASIC` or `CLIENT_SECRET_POST` when a client is registered with `BASIC` or `POST` respectively, or
2. Automatically upgrade `BASIC` or `POST` to `CLIENT_SECRET_BASIC` or `CLIENT_SECRET_POST` respectively when a client is registered.

Both options could have an impact. Would you be able to attempt option 2? One approach could be limiting changes to `RegisteredClient.Builder#clientAuthenticationMethod`, `RegisteredClient.Builder#clientAuthenticationMethods` and tests.

**Comment by bibibiu2017** [comment]
I have attempted option 2. Though I could not localize impact to the recommended test files due to assertions that check equality with deprecated `BASIC` and `POST`. One recommendation would be to change overridden  `equals()` method of `ClientAuthenticationMethod`  making `BASIC` or `POST` equivalent to `CLIENT_SECRET_BASIC ` or `CLIENT_SECRET_POST ` respectively.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java [hunk 1]
```diff
@@ -498,8 +490,9 @@ private RegisteredClient create() {
 			registeredClient.clientSecret = this.clientSecret;
 			registeredClient.clientSecretExpiresAt = this.clientSecretExpiresAt;
 			registeredClient.clientName = this.clientName;
-			registeredClient.clientAuthenticationMethods = Collections.unmodifiableSet(
-					new HashSet<>(this.clientAuthenticationMethods));
+			registeredClient.clientAuthenticationMethods = this.clientAuthenticationMethods.stream()
+					.map(this::upgradeDeprecatedAuthenticationMethod)
+					.collect(Collectors.collectingAndThen(Collectors.toCollection(HashSet::new), Collections::unmodifiableSet));
```
Comment (sjohnr) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java:495]: This is really nice! Unfortunately, we will want to [avoid the use of streams](https://github.com/spring-projects/spring-security/issues/9642).

Perhaps a new private method called from inside `build()` similar to `validateScopes()` and `validateRedirectUris()`:

```java
		private void upgradeClientAuthenticationMethods() {
			if (this.clientAuthenticationMethods.contains(ClientAuthenticationMethod.BASIC)) {
				this.clientAuthenticationMethods.remove(ClientAuthenticationMethod.BASIC);
				this.clientAuthenticationMethods.add(ClientAuthenticationMethod.CLIENT_SECRET_BASIC);
			}
			if (this.clientAuthenticationMethods.contains(ClientAuthenticationMethod.POST)) {
				this.clientAuthenticationMethods.remove(ClientAuthenticationMethod.POST);
				this.clientAuthenticationMethods.add(ClientAuthenticationMethod.CLIENT_SECRET_POST);
			}
		}
```

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java [hunk 2]
```diff
@@ -15,25 +15,17 @@
...
 
+import org.springframework.lang.Nullable;
+import org.springframework.security.oauth2.core.*;
+import org.springframework.security.oauth2.server.authorization.config.*;
+import org.springframework.util.*;
+
 import java.io.Serializable;
-import java.net.URI;
-import java.net.URISyntaxException;
+import java.net.*;
```
Comment (sjohnr) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java:24]: Your IDE may be too aggressively condensing imports. Wildcards shouldn't be used.

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java [hunk 3]
```diff
@@ -291,7 +291,7 @@ public void buildWhenTwoClientAuthenticationMethodsAreProvidedThenBothAreRegiste
 				.build();
 
 		assertThat(registration.getClientAuthenticationMethods())
-				.containsExactlyInAnyOrder(ClientAuthenticationMethod.BASIC, ClientAuthenticationMethod.POST);
+				.containsExactlyInAnyOrder(ClientAuthenticationMethod.CLIENT_SECRET_BASIC, ClientAuthenticationMethod.CLIENT_SECRET_POST);
```
Comment (sjohnr) [inline oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java:294]: I think we should add an additional test here that uses the non-deprecated methods, but retain this one to test the upgrade process. Perhaps this one could be renamed to indicate it's testing the upgrade. `buildWhenBothDeprecatedClientAuthenticationMethodsAreProvidedThenBothNonDeprecatedAreRegistered`

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java [hunk 4]
```diff
@@ -309,7 +309,7 @@ public void buildWhenClientAuthenticationMethodsConsumerIsProvidedThenConsumerAc
...
```
Comment (sjohnr) [inline oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java:312]: I think we should add an additional test here that uses the non-deprecated methods, but retain this one to test the upgrade process. Perhaps this one could be renamed to indicate it's testing the upgrade. `buildWhenConsumerAddsDeprecatedClientAuthenticationMethodsThenNonDeprecatedAreRegistered`.

**Review by sjohnr — COMMENTED** [review]
@bibibiu2017, sorry I should have been more clear, I meant "limiting **additional** changes outside of tests" and I think you have done that! I was just thinking of a slightly different approach than yours, but they work out to the same thing. So this seems like a great option!

See inline feedback below. Just a note I'm not going to merge this right away, as I'd like to have a discussion with our community maintainer first, but if we can get it ready for merge, then we'll be in really good shape.

**Comment by sjohnr** [comment]
Also, I forgot to mention, make sure your commit message contains on a separate line:

"Closes gh-346"

**Comment by bibibiu2017** [comment]
@sjohnr thanks for the feedback will be working on this shortly.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java (hunk 1 above)
Comment (bibibiu2017) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java]: I have removed the stream api

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java (hunk 2 above)
Comment (bibibiu2017) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java]: I have removed world card imports

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java (hunk 4 above)
Comment (bibibiu2017) [reply oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java:346]: I have added the extra test `buildWhenConsumerAddsDeprecatedClientAuthenticationMethodsThenNonDeprecatedAreRegistered `

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java (hunk 3 above)
Comment (bibibiu2017) [reply oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClientTests.java:310]: I have added the extra test `buildWhenBothDeprecatedClientAuthenticationMethodsAreProvidedThenBothNonDeprecatedAreRegistered `

**Comment by jgrandja** [comment]
Thanks for all the updates @bibibiu2017 ! This is now in main.

I also added a polish commit that ensures all usages of `ClientAuthenticationMethod.BASIC` and `ClientAuthenticationMethod.POST` are no longer used.

**Pull request closed without merging**